### DISCUSSION TITLE
Token Exchange from Confidential to Public Client in v18+

### DISCUSSION THREAD

**Discussion by skhro87** [keycloak/keycloak#14367] [Q&A]
Hello Community,

After trying to solve this issue myself for some time, I kindly ask for your help and guidance. 

We have the following multi-tenant setup in our Keycloak:

- One "Internal" realm for internal support staff.
- N number of "Customer" realms for customer's users (each has a public "frontend" client and a confidential "internal_service" client)
- Default built-in "Master" Keycloak realm

We want the "Internal" users to be able to "login as user" into any "Customer" realm.
We implemented the following flow:
1. "Internal" user gets token from KC 
2. "Internal" user requests to impersonate target "Customer" user to Backend
3. Backend uses secret Master realm credentials to get an access token for the target realm's "internal_service" confidential client
4. Backend uses the acquired access token of step 3 to perform a token-exchange to the target user and target "frontend" client
5. Backend returns target user's access token to the initial requester of step 1.

The flow used to work before upgrading to Keycloak 18, however it is now broken.
When performing the token exchange in step 4, the response is:
```
{
    "error": "access_denied",
    "error_description": "Client is not the holder of the token"
}
```

From some research, I understand that the problem is that the `subject_token` in the token-exchange has a different `azp` ("internal_service") then then `client_id` that we perform the token-exchange for ("frontend").

Is there a better way to perform the same functionality? 
Am I overlooking something? I tried to find ways to overwrite the `azp` when getting the "internal_service" confidential client token in step 3, but this does not seem possible (which makes sense, from security perspective).

Any help would be greatly appreciated, thank you.

**Comment by pazaan** [comment]
We recently came across a similar thing when upgrading from 16 to 20.
We resolved it by enabling `admin-fine-grained-authz`, turning on `Fine grained permissions` for our target client, and creating the `token-exchange` permission for the target client. Then we created a Client Policy for the source client, and applied that new policy to the `token-exchange` permission we created for the target client.

**Reply by skhro87** [reply]
thank you! I will try that out!

**Reply by IshwarSingh16** [reply]
Hi @skhro87 did u get any solution?
